cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -1091,36 +1089,97 @@ private ArtifactCell(Id.Artifact artifactId) {
 
   // Data that will be stored for an artifact. Same as ArtifactDetail, expected without the id since that is redundant.
   private static class ArtifactData {
+    // URI for Backward Compatibility
     private final URI locationURI;
+    private final String locationPath;
     private final ArtifactMeta meta;
 
     ArtifactData(Location location, ArtifactMeta meta) {
-      this.locationURI = location.toURI();
+      this(Locations.getRelativePath(location), meta);
+    }
+
+    ArtifactData(String locationPath, ArtifactMeta meta) {
+      this.locationURI = null;
+      this.locationPath = locationPath;
+      this.meta = meta;
+    }
+
+    ArtifactData(URI locationURI, ArtifactMeta meta) {
+      this.locationURI = locationURI;
+      this.locationPath = null;
       this.meta = meta;
     }
+
+    private Location getLocation(LocationFactory locationFactory) {
+      Location artifactLocation = this.locationURI != null
+        ? locationFactory.create(Locations.getRelativePath(locationFactory.create(this.locationURI)))
+        : locationFactory.create(this.locationPath);
+      return artifactLocation;
+    }
   }
 
   // Data that will be stored for a plugin.
   private static class PluginData {
     private final PluginClass pluginClass;
     private final ArtifactRange usableBy;
+    // URI for Backward Compatibility
     private final URI artifactLocationURI;
+    private final String artifactLocationPath;
 
     PluginData(PluginClass pluginClass, ArtifactRange usableBy, Location artifactLocation) {
+      this(pluginClass, usableBy, Locations.getRelativePath(artifactLocation));
+    }
+
+    PluginData(PluginClass pluginClass, ArtifactRange usableBy, String artifactLocationPath) {
+      this.pluginClass = pluginClass;
+      this.usableBy = usableBy;
+      this.artifactLocationURI = null;
+      this.artifactLocationPath = artifactLocationPath;
+    }
+
+    PluginData(PluginClass pluginClass, ArtifactRange usableBy, URI artifactLocationURI) {
       this.pluginClass = pluginClass;
       this.usableBy = usableBy;
-      this.artifactLocationURI = artifactLocation.toURI();
+      this.artifactLocationURI = artifactLocationURI;
+      this.artifactLocationPath = null;
+    }
+
+    private Location getLocation(LocationFactory locationFactory) {
+      Location artifactLocation = this.artifactLocationURI != null
```

Comment:
this logic is repeated 3 times and will be used in other places too. We can move it to the Locations class as something like `Locations.getCompatibleLocation(LocationFactory locationFactory, @Nullable String path, @Nullable URI uri)`